Added amazon-connect-streams from v1.3

andyhopp opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

andyhopp commented:
Forgot to check the boxes; fixed.

sheetalkamat commented on types/amazon-connect-streams/index.d.ts:
```diff
@@ -0,0 +1,688 @@
+// Type definitions for non-npm package Amazon Connect Streams API 1.3
+// Project: amazon-connect-streams
```
This needs to be link to the project without that we cannot verify if definitions actually match the structure you are defining here.

andyhopp replied:
I have amended the .d.ts Project header to include a link to the amazon-connect-streams project.

sheetalkamat commented on types/amazon-connect-streams/tsconfig.json:
```diff
@@ -0,0 +1,24 @@
+        ],
+        "noImplicitAny": true,
+        "noImplicitThis": true,
+        "strictNullChecks": false,
```
All new package definitions need to enable `strictNullChecks` so please enable and fix any errors.

andyhopp replied:
Sorry! I had disabled it for the test script. Re-enabled and tweaked the test file.

sheetalkamat commented on types/amazon-connect-streams/index.d.ts:
```diff
@@ -0,0 +1,688 @@
+// Type definitions for non-npm package Amazon Connect Streams API 1.3
+// Project: amazon-connect-streams
+// Definitions by: Andy Hopper <https://github.com/andyhopp>
```
There is already definition at https://github.com/aws/amazon-connect-streams/blob/master/src/index.d.ts
Since this is shipping with definition, I don't think you need them here.

andyhopp replied:
There isn't an NPM package for this library yet; I'm the author of that definition (used my work account) and have added it to DT in order to make it easier for developers

sheetalkamat replied:
Since the definitions are available where the source is, I don't think we need definitions on DT. https://github.com/aws/amazon-connect-streams#downloading-streams shows the steps to get the js files, .d.ts file should be available along with that?

andyhopp replied:
My chief concern is that, since we do not yet publish amazon-connect-streams as an NPM package, there are a couple of manual steps the developer would need to perform to modify their tsconfig to import the .d.ts file. Using the already-idiomatic @types scope simplifies the developer experience and is already a familiar process for developers using external JS libraries.
 
Once we start publishing ACS to NPM, we would, of course, remove the DT package.
 
Thanks again!

sheetalkamat replied:
Adding @RyanCavanaugh and @sandersn for their opinion on hosting the types while they are still available elsewhere.

andyhopp replied:
I should add; if having the types in the ACS GH repo is the issue, I could certainly remove them (I'd prefer *not* to, of course, but I'm willing to be persuaded otherwise!). I'm just trying to provide developers with a simple/easy way to code against ACS with Typescript while I drum up support for an official NPM package.

sheetalkamat replied:
@andyhop making typings available easily seems to make sense. We see that your steps for install are to clone the repro, build and then copy js. Why not copy `.d.ts` as well next to the .js file? When the package is ready to ship, you can add `typings` field to package json pointing to the `.d.ts`

andyhopp replied:
My concern there is that this becomes an additional step the developer needs to perform (modifying the tsconfig file to add the types). Most TS developers are familiar with the default behavior (to wit: including node_modules/@types), but once you start hacking the types property the developer must explicitly include that as well as the other types.

I certainly could work around this by instructing the developer to manually create a @types/amazon-connect-streams sub-folder, but I feel like that's an abuse.

It sounds like we're over-indexing on the fact that I included the types in the JS repo (which is officially only the JS library). I'll work with that team to address the fact that it's not published to NPM, but in the meantime, I really would like to keep the number of manual steps they need to perform to a minimum. I'm currently unhappy with the fact that users have to run a makefile to create the .js file for the library (again, working with the team to resolve that), and I'd be even unhappier with them having to know how to tweak up their tsconfig file (speaking from experience: the less you touch that file, the better).

Would deleting the index.d.ts file from the amazon-connect-streams repo resolve this to everyone's satisfaction?

sheetalkamat replied:
I think the main concern is that we don't understand the extra step you are talking about. Why is copying the .d.ts file next to `.js` file while you are copying it difficult compared to installing types package?

andyhopp replied:
Because if the developer is working with something like React or Angular, the location of the external javascript libraries is not typically under the project folder created by "ng new" or "create-react-app" and will therefore require modifying the tsconfig file to pull that file in. Performing an "npm i @types/xxx" is something I can expect most TS developers to be familiar with.

andyhopp commented:
Hello! I believe I've addressed the reviewer comments. Since this package is not yet available on NPM, I wanted to give site developers a way to integrate the typings through the @types scope. I've updated the header to point to the GitHub repo for angular-connect-streams and resolved the strict null checks issue by refactoring the test script.

andyhopp commented:
@sheetalkamat Sorry, didn't "@" you... Please see above (and thanks for reviewing!).

sheetalkamat commented on types/amazon-connect-streams/index.d.ts:
```diff
@@ -0,0 +1,688 @@
+        softphone?: SoftPhoneOptions;
+     }
+
+     enum AgentStateType {
```
Is this really enum? Do the values like `AgentStateType` or `AgentAvailStates` etc exist at runtime. If not you want to declare this as union of string literal instead as :
```ts
type AgentStateType  = "init" | "routable" | "not_routable" | "offline";
```

andyhopp replied:
Hello! Yep, it is an enum; the expected usage per the ACS API docs is AgentStateType.INIT (ACS creates an enum object using the connect.makeEnum utility function).

andyhopp commented:
Thanks to everyone!!!
